staging/kos/pkg/controllers/connectionagent/connection_agent.go

```diff
@@ -0,0 +1,1180 @@
+		ifcOwnerNSN := parse.AttNSN(ifcOwner)
+		_, ownerAlreadyHasInterface := ca.getNetworkInterface(ifcOwnerNSN)
+
+		// Pre-existing Network interfaces are matched to network attachments
+		// on the basis of the VNI, the IP and the host of the attachment. But
+		// these fields can be seen changing even if the namespaced name is
+		// steady, for instance by deleting the network attachment and creating
+		// one with same namespaced name and a different VNI. This means that
+		// two network interfaces could be matched to the same namespaced name,
+		// where each match would correspond to two different versions of the
+		// same network attachment, or two different network attachments with
+		// the same namespaced name. In such cases, the interface to keep is the
+		// one matching the most recent version of the network attachment, while
+		// the other should be deleted. Notice that the order in which the
+		// matches took place does not say which match should be kept, because
+		// the two versions of the matched network attachment might come from
+		// different informers and there are no cross-informer ordering
+		// guarantees. There are ways to always take the optimal choice, but
+		// they make for really complex code and yield little advantage because
+		// attachments can change: if an attachment is matched and it changes 1
+		// sec later, the match was useless. For this reason, this code does not
+		// attempt to always take the optimal choice: in cases of collisions the
+		// interface that was matched first is taken and the other is deleted.
+		// Collisions should be a rare event anyway. Even if the wrong choice is
+		// taken during normal operation the connection agent will rectify the
+		// the mistake by realizing the network interface does not match the
+		// attachment and creating a correct interface after deleting the old one.
+		if !ownerAlreadyHasInterface {
+			ca.assignNetworkInterface(ifcOwnerNSN, ifc)
+			klog.V(3).Infof("Matched pre-existing network interface %s with attachment %s", ifc, ifcOwnerNSN)
+			if localIfc, ifcIsLocal := ifc.(*localNetworkInterface); ifcIsLocal {
+				ca.setExecReport(ifcOwnerNSN, localIfc.id, ifcOwner.Status.PostCreateExecReport)
+				localIfc.postDeleteExec = ifcOwner.Spec.PostDeleteExec
+			} else {
+				// The interface is remote and so is its owner. There's no
+				// guarantee that the add notification handler for the owner has
+				// executed yet, neither as for when it will execute. If that
+				// has not happened by the time a worker processes the deletion
+				// of the last local NetworkAttachment in the owner's virtual
+				// network, the layer 1 of the virtual network state is cleared
+				// before the owner's namespaced name could be recorded in it.
+				// This means the owner is not processed, hence its network
+				// interface is not deleted even if it should. To avoid this,
+				// enqueue the owner to force its processing.
+				ca.queue.Add(ifcOwnerNSN)
```

Comment:
Enqueuing can be inefficient because if the virtual network stays relevant and workers process `ifcOwner` before its add notification handler adds it to layer 1 of the virtual network state, `ifcOwner` appears to have been deleted even if it isn't and its network interface is deleted. This is recovered as soon as the add notification handler executes and a worker re-processes `ifcOwner`, but it's a waste of time.

However, notice that there's an alternative to enqueuing: adding the `ifcOwner` to layer 1 of the virtual network state. That can be either achieved with a new, ad-hoc function, or by re-using https://github.com/MikeSpreitzer/kube-examples/pull/81/files#L487 . The latter requires a few tweaks, because that function has `vnStateUID` as a parameter which, in this specific case, is unneeded and the client does not know which value to pass in. One solution is to define as reserved value for `vnStateUID`  to use only in this case. vnStateUIDs are generated from the same variable which is always increased. If we init such variable to 1 we can safely use 0 as the reserved value.

My net net was that enqueuing was the simpler solution and that its drawbacks happen only under rare circumstances (this is only a guess though),  that is, by the time a worker dequeues `ifcOwnerNSN` it's very likely that `ifcOwnerNSN` has been already added to its virtual network's layer 1 state by the add notification handler.

Reply:
The source of the problem is that we think an informer can claim "sync" after its cache is populated but before all the corresponding notification calls have completed, right?  I think a simple way to handle that may be to make `syncPreExistingNetworkInterfaces` do duplicate notification calls (synchronously, so we know they complete).  The normal create notifications will all precede corresponding delete notifications (informers promise us that).  We launch the workers only after the notifications in `syncPreExistingNetworkInterfaces` have completed.

Let's not manage VN state IDs if we do not have to.  I think we can just use the `*layer1VirtualNetworkState` itself instead.

Reply:
> Let's not manage VN state IDs if we do not have to. I think we can just use the *layer1VirtualNetworkState itself instead.

I'm not too fond of UIDs either, but I see one difficulty with what you're proposing. Part of Layer 1 state is per-virtual network (`layer1VirtualNetworkState`, most importantly `remoteAtts`), while part of it is global (`ca.l1VirtNetsState`, most importantly `attToSeenInCaches`). If an add notification handler executes after its virtual network L1 state was cleared due to becoming irrelevant and after the same virtual network became relevant again, the handler must NOT update the new L1 state, either global (`attToSeenInCaches`) or per-virtual network (`remoteAtts`). UIDs make that possible, what you're proposing ensures that the per-virtual network L1 state is not touched, but offers no way to tell whether `attToSeenInCaches` should be updated or not.

Reply:
My previous comment assumed you meant to use the *layer1VirtualNetworkState passed to the notification handler only to directly update it. I didn't realize you can also use it as an implicit UID by comparing it against the one currently stored in `ca.l1VirtNetsState.vniToVNState[vni]`. That way you can avoid updating `attToSeenInCaches` as well.

Reply:
I implemented what's described in my previous comment in 45701a6.